pelican/plugins/seo/seo.py

```diff
@@ -43,28 +36,67 @@ def plugin_initializer(settings):
     logger.info("SEO plugin initialized")
 
 
+def get_plugin_settings(context):
+    """Get settings in the pelican config file from the given context
+
+    .. note:: Pelican's settings take precedence over the plugin's default settings.
+
+    SEO report settings:
+
+        - SEO_REPORT
+        - SEO_ARTICLES_LIMIT
+        - SEO_PAGES_LIMIT
+
+    SEO enhancer settings:
+
+        - SEO_ENHANCER
+        - SEO_ENHANCER_OPEN_GRAPH
+        - SEO_ENHANCER_TWITTER_CARDS
+
+    :return: Dictionary of settings for the plugin
+    :rtype: <dict>
+    """
+
+    def get_seo_variables(data):
+        """Get variables starting with 'SEO_' & values in the given dictionary
+        :rtype: <dict>
+        """
+        return {
+            var_name: value
+            for var_name, value in data.items()
+            if var_name.startswith("SEO_")
+        }
+
+    plugin_settings = get_seo_variables(vars(settings))
```

Comment:
> So if I understand well, you populate the plugin_settings dict with the plugin settings first, then you add settings from Pelican. Is that right?

Absolutely, Pelican's settings (in the files configured by the users) have priority over the module settings.

> As we want to only get settings from plugin only if there is no settings from Pelican

This is exactly what the code does. This way you can be sure that what the function contains all the plugin-specific variables that the plugin might need.

With the added assurance that these variables are updated by the user's choices.
This is exactly how a default variable is overwritten by a specific choice (argparse, etc.).

Moreover, during a future maintenance, you won't have to modify this code, just the docstring, the global readme and the settings.py file (which is already quite enough).


I am not in favor of the alternative you suggest because:

- it uses an import included in a function:
not recommended except in rare cases of circular imports or very heavy modules
- requires more maintenance work in the future
- does not guarantee that all necessary and sufficient variables are in the dictionary obtained from userspace.


I recognize that the call to `get_plugin_settings()` can look heavy because `run_html_enhancer()`  calls it for each file processed (so for each blog post).

However, the cost of this call is infinitesimal compared to the I/O work done by the other
other functions of the plugin;
moreover with the resolution of issue #55 (which I hope :grin:), the `run_html_enhancer()` function function would be called only once, that is to say 3 times in the whole program.

However, we could indeed consider caching once and for all the return from `get_plugin_settings()` by adding a key/flag to the `context` dictionary, common to all Pelican objects.

If this flag is present, then the parameters have been processed and can be safely read
from `generator.context` without having to call `get_plugin_settings()`.
To my knowledge no plugin does this, it looks like premature optimization but, why not.

For test purpose:

```python
def get_plugin_settings(context):
    def get_seo_variables(settings):
        """Get variables starting with 'SEO_' & values in the given dictionary
        :rtype: <dict>
        """
        return {
            name: value
            for name, value in settings.items()
            if name.startswith("SEO_")
        }

    plugin_settings = get_seo_variables(default_plugin_settings)
    plugin_settings.update(get_seo_variables(context))
    return plugin_settings


default_plugin_settings = {'SEO_REPORT': True, 'SEO_ENHANCER': False, 'SEO_ENHANCER_OPEN_GRAPH': False,
'SEO_ENHANCER_TWITTER_CARDS': False, 'SEO_ARTICLES_LIMIT': 10, 'SEO_PAGES_LIMIT': 10}
# For more realism add some garbage keys
default_plugin_settings.update(globals())

context = {'SEO_REPORT': True, 'SEO_ENHANCER': True}

import timeit
from functools import partial
my_func = partial(get_plugin_settings, context)
print(timeit.timeit(stmt=my_func, number=1000000))
```

Result is 5.84 sec for 1 million of calls. so 5.84 microsecond per call.

No risk that the slightest optimization can amortize the time I spent writing this note :stuck_out_tongue_closed_eyes:

Reply:
Even if we don't go for the slightest optimization, be sure that you didn't loose your time explaining this, because you taught me many things and it's very valuable to me :)

I still need to evaluate and do some tests about #55, but if this change will avoid calling the `get_plugin_settings` for each page / article, it's always better, even if it's an infinitesimal cost.

I think that at term, `settings.py` should be deprecated and deleted.